external/storm-kafka-new-consumer-api/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,445 @@
+
+
+    public KafkaSpout(KafkaSpoutConfig<K,V> kafkaSpoutConfig, KafkaSpoutStreams kafkaSpoutStreams, KafkaTupleBuilder<K,V> tupleBuilder) {
+        this.kafkaSpoutConfig = kafkaSpoutConfig;                 // Pass in configuration
+        this.kafkaSpoutStreams = kafkaSpoutStreams;
+        this.tupleBuilder = tupleBuilder;
+    }
+
+    @Override
+    public void open(Map conf, TopologyContext context, SpoutOutputCollector collector) {
+        initialized = false;
+
+        // Spout internals
+        this.collector = collector;
+        maxRetries = kafkaSpoutConfig.getMaxTupleRetries();
+
+        // Offset management
+        firstPollOffsetStrategy = kafkaSpoutConfig.getFirstPollOffsetStrategy();
+        consumerAutoCommitMode = kafkaSpoutConfig.isConsumerAutoCommitMode();
+
+        if (!consumerAutoCommitMode) {     // If it is auto commit, no need to commit offsets manually
+            timer = new Timer(kafkaSpoutConfig.getOffsetsCommitFreqMs(), 500, TimeUnit.MILLISECONDS);
+            acked = new HashMap<>();
+        }
+
+        // Kafka consumer
+        kafkaConsumer = new KafkaConsumer<>(kafkaSpoutConfig.getKafkaProps(),
+                kafkaSpoutConfig.getKeyDeserializer(), kafkaSpoutConfig.getValueDeserializer());
+        kafkaConsumer.subscribe(kafkaSpoutConfig.getSubscribedTopics(), new KafkaSpoutConsumerRebalanceListener());
+        // Initial poll to get the consumer registration process going.
+        // KafkaSpoutConsumerRebalanceListener will be called foloowing this poll upon partition registration
+        kafkaConsumer.poll(0);
+
+        LOG.debug("Kafka Spout opened with the following configuration: {}", kafkaSpoutConfig.toString());
+    }
+
+    // ======== Next Tuple =======
+
+    @Override
+    public void nextTuple() {
+        if (initialized) {
+            if(commit()) {
+                commitOffsetsForAckedTuples();
+            } else {
+                emitTuples(poll());
+            }
+        } else {
+            LOG.debug("Spout not initialized. Not sending tuples until initialization completes");
+        }
+    }
+
+    private boolean commit() {
+        return !consumerAutoCommitMode && timer.expired();    // timer != null for non auto commit mode
+    }
+
+    private ConsumerRecords<K, V> poll() {
+        final ConsumerRecords<K, V> consumerRecords = kafkaConsumer.poll(kafkaSpoutConfig.getPollTimeoutMs());
+        LOG.debug("Polled [{}] records from Kafka", consumerRecords.count());
+        return consumerRecords;
+    }
+
+    private void emitTuples(ConsumerRecords<K, V> consumerRecords) {
+        for (TopicPartition tp : consumerRecords.partitions()) {
+            final Iterable<ConsumerRecord<K, V>> records = consumerRecords.records(tp.topic());     // TODO Decide if want to give flexibility to emmit/poll either per topic or per partition
```

Comment:
TODOs should not be left in the code.  We should either do this, skip it, or file a follow on JIRA for it.  Either way we should delete the comment. I personally am fine either way, but releasing a spout sooner rather then later seems better to me.